neuralmonkey/learning_utils.py

```diff
@@ -156,48 +184,53 @@ def training_loop(tf_manager: TensorFlowManager,
-                        .format(epoch_n, batch_n), color='blue')
-
-                    _log_continuous_evaluation(tb_writer, tf_manager,
-                                               main_metric,
-                                               val_evaluation,
-                                               seen_instances, epoch_n,
-                                               epochs,
-                                               val_results, train=False)
-
-                    if this_score == tf_manager.best_score:
-                        best_score_str = colored(
-                            "{:.4g}".format(tf_manager.best_score),
-                            attrs=['bold'])
-                    else:
-                        best_score_str = "{:.4g}".format(tf_manager.best_score)
-
-                    log("best {} on validation: {} (in epoch {}, "
-                        "after batch number {})"
-                        .format(main_metric, best_score_str,
-                                tf_manager.best_score_epoch,
-                                tf_manager.best_score_batch),
-                        color='blue')
-
-                    log_print("")
-                    _print_examples(val_dataset, val_outputs,
-                                    val_preview_input_series,
-                                    val_preview_output_series,
-                                    val_preview_num_examples)
+                    for validation_dataset in val_datasets:
+                        val_results, val_outputs = run_on_dataset(
+                            tf_manager, runners, validation_dataset,
+                            postprocess, write_out=False,
+                            batch_size=runners_batch_size)
+                        # ensure val outputs are iterable more than once
+                        val_outputs = {k: list(v)
+                                       for k, v in val_outputs.items()}
+                        val_evaluation = evaluation(
+                            evaluators, validation_dataset, runners,
+                            val_results,
+                            val_outputs)
+
+                        this_score = val_evaluation[main_metric]
+                        tf_manager.validation_hook(this_score, epoch_n,
+                                                   batch_n)
+
+                        log("Validation (epoch {}, batch number {}):"
+                            .format(epoch_n, batch_n), color='blue')
+
+                        _log_continuous_evaluation(tb_writer, tf_manager,
+                                                   main_metric,
+                                                   val_evaluation,
+                                                   seen_instances, epoch_n,
+                                                   epochs,
+                                                   val_results, train=False)
+
+                        if this_score == tf_manager.best_score:
```

Comment:
Takhle, jak to tady máš napsaný, by to znamenalo, že když se narazí na numericky nejlepší hodnotu na libovolném z validačních datasetů, tak se to zapamatuje jako nejlepší skóre. Ale třeba BLEU skóre není mezi různými datasety vůbec porovnatelné, u jiných metrik můžou být datasety různě obtížné. Osobně bych se klonil k tomu, aby se rozhodovalo jenom podle jednoho validačního datasetu, kdy se uloží proměnné, a kdy je model nejlepší, protože jinak to nejde jednoznačně definovat.